tripy/examples/resent50/tripy-res50.ipynb

```diff
@@ -0,0 +1,599 @@
+    "        for idx, (num_layers, out_channels, bottleneck_channels, stride) in enumerate(layers_config):\n",
+    "            stage = ResNetStage(num_layers, in_channels, out_channels, bottleneck_channels, stride)\n",
+    "            self.stages.append(stage)\n",
+    "            in_channels = out_channels\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        for stage in self.stages:\n",
+    "            x = stage(x)\n",
+    "        return x\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define a max pooling class and embeddings class for input pre-procDefine a max pooling class and embeddings class for input pre-processing before feeding to encoderessing before feeding to encoder"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MaxPool2d(tp.Module):\n",
+    "    def __init__(self, kernel_dims, stride=None, padding=((0, 0), (0, 0))):\n",
+    "        super(MaxPool2d, self).__init__()\n",
+    "        self.kernel_dims = kernel_dims\n",
+    "        self.stride = stride if stride is not None else kernel_dims\n",
+    "        self.padding = padding\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        return tp.maxpool(x, kernel_dims=self.kernel_dims, stride=self.stride, padding=self.padding)\n",
+    "\n",
+    "\n",
+    "class ResNetEmbeddings(tp.Module):\n",
+    "    def __init__(self):\n",
+    "        super(ResNetEmbeddings, self).__init__()\n",
+    "        self.embedder = ResNetConvLayer(3, 64, kernel_dims=(7, 7), stride=(2, 2), padding=((3, 3), (3, 3)))\n",
+    "        self.pooler = MaxPool2d(kernel_dims=(3, 3), stride=(2, 2), padding=((1, 1), (1, 1)))\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        x = self.embedder(x)\n",
+    "        x = self.pooler(x)\n",
+    "        return x"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define a custom adaptive average pooling layer and the resenet backbone model itself"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class AdaptiveAvgPool2d(tp.Module):\n",
+    "    def __init__(self, output_size):\n",
+    "        super(AdaptiveAvgPool2d, self).__init__()\n",
+    "        self.output_size = output_size\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        N, C, H_in, W_in = x.shape\n",
+    "        H_out, W_out = self.output_size\n",
+    "\n",
+    "        # Calculate stride and kernel size\n",
+    "        stride_h = H_in // H_out\n",
+    "        stride_w = W_in // W_out\n",
+    "\n",
+    "        kernel_size_h = H_in - (H_out - 1) * stride_h\n",
+    "        kernel_size_w = W_in - (W_out - 1) * stride_w\n",
+    "\n",
+    "        return tp.avgpool(x, kernel_dims=(int(kernel_size_h), int(kernel_size_w)), stride=(int(stride_h), int(stride_w)))\n",
```

Comment:
Could we just make this a helper function instead of a module?

Reply:
done